File: courses/CS50X/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // Promt for Input
    long x = get_long("Card Number: ");
    long s = x;

    // calculate cecksum "p" and check for length "n"
    int a;
    int y;
    int z;
    int p;
    int n;
    y = 0;
    z = 0;
    for (n = 0; x > 0; n++)
    {
        if (n % 2 == 0)
        {
            a = 0;
            a = (x % 10);
            y = y + (a % 10);
            x = (x - a) / 10;
        }
        if (n % 2 == 1)
        {
            a = 0;
            a = (x % 10);
            z = z + ((2 * a) % 10) + ((2 * a) / 10);
            x = (x - a) / 10;
        }
    }
    p = 0;
    p = z + y;

    // check for starting digits "s"
    long t;
    t = pow(10, (n - 2));
    s = s / t;

    // Print results
    if (p % 10 == 0)
    {
        if (n == 15 && (s == 34 || s == 37))
        {
            printf("AMEX\n");
        }
        else if (n == 16 && (s > 50 && s < 56))
        {
            printf("MASTERCARD\n");
        }
        else if ((n == 13 || n == 16) && s / 10 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: courses/CS50X/sentimental-credit/credit.py
c# TODO


def main():
    while True:
        crdnbr = promt_user()
        if crdnbr.strip("0123456789") == "":
            break

    if check_luhn(crdnbr) == True:
        card_type(crdnbr)
    else:
        print("INVALID")


def promt_user():
    crdnbr = input("Number: ")
    return crdnbr


def check_luhn(crdnbr):
    n = len(crdnbr) - 1
    sum = 0

    while n >= 0:
        sum += int(crdnbr[n])
        n -= 1
        if n >= 0:
            pd = int(crdnbr[n]) * 2
            sum += (pd // 10) + (pd % 10)
            n -= 1

    if sum % 10 == 0:
        return True

    return False


def card_type(crdnbr):
    match crdnbr[:2]:
        case "34" | "37":
            if len(crdnbr) == 15:
                print("AMEX")
                return
        case "51" | "52" | "53" | "54" | "55":
            if len(crdnbr) == 16:
                print("MASTERCARD")
                return

    if crdnbr[0] == "4" and (len(crdnbr) == 13 or len(crdnbr) == 16):
        print("VISA")
        return

    print("INVALID")


if __name__ == "__main__":
    main()
